services/src/main/java/org/keycloak/services/clientpolicy/ClientUpdateContext.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.services.clientpolicy;
+
+import org.keycloak.models.ClientModel;
+import org.keycloak.representations.idm.ClientRepresentation;
+import org.keycloak.services.clientregistration.ClientRegistrationContext;
+import org.keycloak.services.clientregistration.policy.RegistrationAuth;
+import org.keycloak.services.resources.admin.AdminAuth;
+
+/**
+ * Represents the context in the client registration/update by Dynamic Client Registration or Admin REST API.
+ */
+public interface ClientUpdateContext extends ClientPolicyContext {
+
+    /**
+     * returns {@link ClientModel} of the current client that will be updated.
+     *
+     * @return {@link ClientModel}
+     */
+    default ClientModel getCurrentClientModel() {return null;}
+
+    /**
+     * returns {@link ClientRepresentation} for updating the current client by Admin REST API.
+     *
+     * @return {@link ClientRepresentation}
+     */
+    default ClientRepresentation getProposedClientRepresentation() {return null;}
+
+    /**
+     * returns {@link RegistrationAuth} by Dynamic Client Registration.
+     *
+     * @return {@link RegistrationAuth}
+     */
+    default RegistrationAuth getDynamicRegistrationAuth() {return null;}
+
+    /**
+     * returns {@link AdminAuth} by Admin REST API.
+     *
+     * @return {@link AdminAuth}
+     */
+    default AdminAuth getAdminAuth() {return null;}
```

Comment:
I would prefer if there's a single way to get auth context rather than two separate depending on what API is used. That is enhancements/polish we can do later though.

Reply:
I see, I'll use the same methods for Dynamic Client Registration and Admin REST API.

### Creating/Updating Client
- `ClientRepresentation getProposedClientRepresentation()`
  `ClientRepresentation` for creating or updating the current client.
- `ClientModel getClientToBeUpdated()`
  `ClientModel` that will be updated.

### Authentication Context
- `UserModel getAuthenticatedUser()`
  The authenticated and authorized user for creating /updating a client. There is the case that no such user exist. (e.g. creating/updating a client with an access token issued by Client Credentials Grant)
- `ClientModel getAuthenticatedClient()`
  The authenticated and authorized client for creating /updating a client. There is the case that no such client exist. (e.g. creating/updating a client with an initial access token or a registration access token, or without any token)
- `JsonWebToken getToken()`
  The token accompanied with a registration/update client request. It will be used to do some policy operation based on it in the future.

WDYT?